computer_vision/face_recognition_using_eigenfaces.py

```diff
@@ -0,0 +1,454 @@
+Link to the research paper: https://ieeexplore.ieee.org/document/139758 <br>
+Google Colab Notebook: https://colab.research.google.com/drive/1InXv7hbjSBRkuLZgpf9SCFQcxu2m924B#scrollTo=Hj2uRB2ke4Vy
+"""
+
+!pip install opencv-python
+
+# taken from this StackOverflow answer: https://stackoverflow.com/a/39225039
+import requests
+
+
+def download_file_from_google_drive(id, destination):
+    URL = "https://docs.google.com/uc?export=download"
+
+    session = requests.Session()
+
+    response = session.get(URL, params = { 'id' : id }, stream = True)
+    token = get_confirm_token(response)
+
+    if token:
+        params = { 'id' : id, 'confirm' : token }
+        response = session.get(URL, params = params, stream = True)
+
+    save_response_content(response, destination)
+
+def get_confirm_token(response):
+    for key, value in response.cookies.items():
+        if key.startswith('download_warning'):
+            return value
+
+    return None
+
+def save_response_content(response, destination):
+    CHUNK_SIZE = 32768
+
+    with open(destination, "wb") as f:
+        for chunk in response.iter_content(CHUNK_SIZE):
+            if chunk: # filter out keep-alive new chunks
+                f.write(chunk)
+
+"""The dataset used was the AT&T dataset, where 10 different images of 40 distinct subjects were given. The 10 images of the same person were taken under varying circumstances, like different lighting, facial expressions and facial accessories. <br>
+
+We took two images of each of the 40 people for testing. The model was trained on 8 images of the first 20 people. We did this so that we can calculate the accuracy for people both inside and outside our training set. <br>
+
+On running the code, train_small.zip, test1.zip and test2.zip will appear in the contents folder. These signify the following:
+<ol>
+<li>train_small: training dataset, contains 8 images of 20 different people.</li><li>test1.zip: testing dataset, contains 2 images all people from the training dataset.</li>
+<li>test2.zip: testing dataset, consisting of 2 images of 20 new people whose images weren't present in the training dataset.</li>
+
+"""
+
+# Downloading training dataset
+file_id = '1Z9evWwUK4nTpARz67ZWraNe5sX2MNcFx'
+destination = '/content/dataset.zip'
+download_file_from_google_drive(file_id, destination)
+
+!unzip -q dataset.zip
+!rm -rf dataset.zip
+
+# Downloading test1 dataset
+test_file_id_1 = '1bIr_ikTxGuuZnJulykXeZz04pyb27S3f'
+test_destination_1 = '/content/test1.zip'
+download_file_from_google_drive(test_file_id_1, test_destination_1)
+
+!unzip -q test1.zip
+!rm -rf test1.zip
+
+# Downloading test2 dataset
+test_file_id_2 = '1Kj0QsxJ1m0n0UCPiaFskskUhM7fvxqS4'
+test_destination_2 = '/content/test2.zip'
+download_file_from_google_drive(test_file_id_2, test_destination_2)
+
+!unzip -q test2.zip
+!rm -rf test2.zip
+
```

Comment:
Perhaps try to make a generalised algorithm which can be applied using any value instead of using those specific assets.